gmeta/src/lib.rs

```diff
@@ -1,66 +1,189 @@
+//!     if let PingPong::Ping = payload {
+//!         unsafe { COUNTER += 1 };
+//!         // Send a reply message of `PingPong` type back to the sender.
+//!         msg::reply(PingPong::Pong, 0).expect("Unable to reply");
+//!     }
+//! }
+//!
+//! #[no_mangle]
+//! extern "C" fn state() {
+//!     msg::reply(unsafe { COUNTER }, 0).expect("Unable to reply");
+//! }
+//! ```
+//!
+//! `build.rs` file:
+//!
+//! ```no_run
+//! # const IGNORE: &'static str = stringify! {
+//! use ping::ContractMetadata;
+//! # };
+//! # pub struct ContractMetadata;
+//! # impl gmeta::Metadata for ContractMetadata {
+//! #     type Init = ();
+//! #     type Handle = ();
+//! #     type Others = ();
+//! #     type Reply = ();
+//! #     type Signal = ();
+//! #     type State = ();
+//! # }
+//!
+//! fn main() {
+//!     gear_wasm_builder::build_with_metadata::<ContractMetadata>();
+//! }
+//! ```
+
 #![no_std]
+#![warn(missing_docs)]
+#![doc(html_logo_url = "https://docs.gear.rs/logo.svg")]
 
 extern crate alloc;
 
 #[cfg(feature = "codegen")]
 pub use gmeta_codegen::metawasm;
 
-pub use alloc::{collections::BTreeMap, string::String};
-pub use scale_info::{
-    scale::{self, Decode, Encode},
-    MetaType, PortableRegistry, Registry, TypeInfo,
-};
+pub use scale_info::{MetaType, PortableRegistry, Registry};
 
-use alloc::vec::Vec;
+use alloc::{collections::BTreeMap, string::String, vec::Vec};
 use blake2_rfc::blake2b;
 use core::any::TypeId;
+use scale_info::{
+    scale::{self, Decode, Encode},
+    TypeInfo,
+};
 
+/// Types representation used by metadata.
 #[derive(Encode, Debug, Decode, Eq, PartialEq)]
 #[codec(crate = scale)]
 pub struct TypesRepr {
+    /// Input types.
     pub input: Option<u32>,
+    /// Output types.
     pub output: Option<u32>,
 }
 
+/// Metadata representation.
 #[derive(Encode, Debug, Decode, Eq, PartialEq)]
 #[codec(crate = scale)]
 pub struct MetadataRepr {
+    /// Init types representation.
     pub init: TypesRepr,
+    /// Handle types representation.
     pub handle: TypesRepr,
+    /// Reply types representation.
     pub reply: TypesRepr,
+    /// Async main types representation.
     pub others: TypesRepr,
+    /// Signal output type representation.
     pub signal: Option<u32>,
+    /// State type representation.
     pub state: Option<u32>,
+    /// Registry.
     pub registry: Vec<u8>,
 }
 
+/// Metawasm data.
 #[derive(Encode, Debug, Decode)]
 #[codec(crate = scale)]
 pub struct MetawasmData {
+    /// Meta functions.
     pub funcs: BTreeMap<String, TypesRepr>,
+    /// Registry.
     pub registry: Vec<u8>,
 }
 
+/// Trait used to get information about types.
 pub trait Type: TypeInfo + 'static {
+    /// Return `true` if type is unit.
     fn is_unit() -> bool {
         TypeId::of::<Self>().eq(&TypeId::of::<()>())
     }
 
+    /// Return [`MetaType`](scale_info::MetaType) information about type.
     fn meta_type() -> MetaType {
         MetaType::new::<Self>()
     }
 
+    /// Register type in registry.
```

Comment:
if not unit, returning its id in registry

Reply:
https://github.com/gear-tech/gear/pull/2761/commits/397596e